### PR TITLE
Change `filter_parameters` to use returned `value` when procs are passed

### PR DESCRIPTION
**Pull request by yui-knk**
Because proc is something like function, it is unnatural to expect procs to
destruct passed `value` to filter parameters.
And ruby has more un-destructive method than destructive method,
so using returned value gives developers more flexibility in filtering
parameters.

### CODE REVIEW

File: actionpack/test/dispatch/request_test.rb [hunk 1]
```diff
@@ -1061,7 +1061,7 @@ class RequestParameterFilter < BaseRequestTest
 
       filter_words << 'blah'
       filter_words << lambda { |key, value|
-        value.reverse! if key =~ /bargain/
```
Comment (rafaelfranca) [inline actionpack/test/dispatch/request_test.rb]: Why this change was needed?

Comment (yui-knk) [inline actionpack/test/dispatch/request_test.rb]: (If "this change" means change from `value.reverse! if key =~ /bargain/` to `key =~ /bargain/ ? value.reverse : value`,) after this commit procs given to `filter_parameters` are expected to return filtered `value` (if filtering is needed) instead of destruct passed `value`.
Although `String#reverse!` returns reversed string, I think it is better to use "String#reverse" because we do only need to **return** filtered `value`.

Comment (rafaelfranca) [inline actionpack/test/dispatch/request_test.rb]: I see. So this means that after this PR applications that have filters that mutate the value will have to change the code?

Comment (yui-knk) [inline actionpack/test/dispatch/request_test.rb]: Yes. Should we warn deprecation first?

Comment (rafaelfranca) [inline actionpack/test/dispatch/request_test.rb]: Yes. If the return of the block is `nil` we warn. Does it make sense?

Comment (yui-knk) [inline actionpack/test/dispatch/request_test.rb]: It's very difficult to judge which is better to warn or not if returned value is `nil`.
When a develop want to filtered value to be changed to `nil`, warnings are noisy. But I the possibility of these use cases are low.
So I agree to warn if the returned value is `nil`, or not String, because before this commit developers can change `value` to be only a String.

Comment (prathamesh-sonpatki) [inline actionpack/test/dispatch/request_test.rb]: > When a develop want to filtered value to be changed to nil, warnings are noisy. But I think the possibility of these use cases are low.

:+1:, I think before this patch someone has to mutate the value because there was no way to get the `filtered` or `original` value back from the block for next filters.

Comment (Schwad) [reply actionpack/test/dispatch/request_test.rb]: @prathamesh-sonpatki that's a fair point
